hdfs3/core.py

```diff
@@ -70,13 +70,22 @@ def conf_to_dict(fname):
 
 def ensure_bytes(s):
     """ Give strings that ctypes is guaranteed to handle """
-    if isinstance(s, dict):
-        return {k: ensure_bytes(v) for k, v in s.items()}
-    if isinstance(s, str) and sys.version_info < (3,):
+    if PY3 and isinstance(s, bytes):
+        return s
+    if not PY3 and isinstance(s, str):
         return s
     if hasattr(s, 'encode'):
         return s.encode()
+    if hasattr(s, 'tobytes'):
+        return s.tobytes()
+    if isinstance(s, bytearray):
+        return bytes(s)
+    if not PY3 and hasattr(s, 'tostring'):
+        return s.tostring()
+    if isinstance(s, dict):
```

Comment:
A coverage report confirms that line is used - not sure yet where, but intend to keep it around.